mm2src/coins/eth/eth_swap_v2.rs

```diff
@@ -0,0 +1,159 @@
+use super::eth::{addr_from_raw_pubkey, gas_limit, wei_from_big_decimal, EthCoin, EthCoinType, SignedEthTx,
+                 TAKER_SWAP_V2};
+use super::{SendTakerFundingArgs, Transaction, TransactionErr};
+use enum_derives::EnumFromStringify;
+use ethabi::Token;
+use ethcore_transaction::Action;
+use ethereum_types::{Address, U256};
+use futures::compat::Future01CompatExt;
+use std::convert::TryInto;
+
+struct TakerFundingArgs<'a> {
+    dex_fee: U256,
+    payment_amount: U256,
+    maker_address: Address,
+    taker_secret_hash: &'a [u8],
+    maker_secret_hash: &'a [u8],
+    funding_time_lock: u32,
+    payment_time_lock: u32,
+}
+
+impl EthCoin {
+    pub(crate) async fn send_taker_funding_impl(
+        &self,
+        args: SendTakerFundingArgs<'_>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let taker_swap_v2_contract = self
+            .swap_v2_contracts
+            .as_ref()
+            .map(|contracts| contracts.taker_swap_v2_contract)
+            .ok_or_else(|| TransactionErr::Plain(ERRL!("Expected swap_v2_contracts to be Some, but found None")))?;
+        let dex_fee = try_tx_s!(wei_from_big_decimal(&args.dex_fee.fee_amount().into(), self.decimals));
```

Comment:
>  Splitting the payment is done in the contract I think

if new swap contract is r2r then we should review it in etomic swap

Reply:
right now we just send dexfee to feeAddress when maker spends payment
https://github.com/KomodoPlatform/etomic-swap/blob/5e15641cbf41766cd5b37b4d71842c270773f788/contracts/EtomicSwapTakerV2.sol#L200

![Screenshot 2024-07-22 at 14 02 31](https://github.com/user-attachments/assets/4392d849-b627-4905-9ef9-8ec7fb77a496)

Reply:
> right now we just send dexfee to feeAddress when maker spends payment

Yep, and `total_spend_amount` will handle this in a right way. When fee splitting is implemented for EVM swaps v2, we won't have to remember changing this part.

Reply:
@shamardy as current swap v2 contracts doesnt have burn functionality, I suggest to leave `dex_fee.fee_amount()`
I will add todo that there is a need of burnFee functionality support.

The preferable approach is in wip I suppose. I would suggest to provide burnFee as an additional param in the swap contract function.

Reply:
> as current swap v2 contracts doesnt have burn functionality, I suggest to leave dex_fee.fee_amount()

As discussed in the call, `dex_fee.total_spend_amount()` returns same value as `dex_fee.fee_amount()` when `DexFee::Standard` is used.

> The preferable approach is in wip I suppose. I would suggest to provide burnFee as an additional param in the swap contract function. And burn fee sending it to zero address right after sending dexFee to feeAddress in spend payment function.

If we will use an additional `burnFee` param, then you can keep using `dex_fee.fee_amount()` but we can't add this `burnFee` param to `ethTakerPayment`, so we should use `dex_fee.total_spend_amount()` for it.

> And burn fee sending it to zero address right after sending dexFee to feeAddress

Why would we send it to zero address?

Reply:
> Why would we send it to zero address?

well burning tokens means sending it to inaccessible address. in evm based chains zero address is commonly used to burn tokens.
https://cointelegraph.com/explained/what-is-the-ethereum-eth-burn-address
https://www.binance.com/en-BH/square/post/8268692784938

Reply:
> dex_fee.total_spend_amount() returns same value as dex_fee.fee_amount() when DexFee::Standard is used.

but if its `DexFee::WithBurn`?

> If we will use an additional burnFee param, then you can keep using dex_fee.fee_amount() but we can't add this burnFee param to ethTakerPayment, so we should use dex_fee.total_spend_amount() for it.

Why? I suggest to add additional burnFee param to `ethTakerPayment` and other swap contract methods.
If `DexFee::Standard` then burnFee will be just zero.

How are you going to split `dexFee` and `burnFee` inside smart contract if you use `total_spend_amount` and use just one `fee` param in smart contract functions?

Reply:
> well burning tokens means sending it to inaccessible address. in evm based chains zero address is commonly used to burn tokens.

As discussed, burnfee will either stay in contract or sent to an address where these fees will be traded for KMD that will be burned.

>  I suggest to add additional burnFee param to ethTakerPayment and other swap contract methods.
If DexFee::Standard then burnFee will be just zero.

Agreed.

> How are you going to split dexFee and burnFee inside smart contract if you use total_spend_amount and use just one fee param in smart contract functions?

As discussed, `total_spend_amount` should be used for `eth_total_payment` https://github.com/KomodoPlatform/komodo-defi-framework/blob/2007a703a6d2cd78c974f5b1fa52de78a9ab190d/mm2src/coins/eth/eth_swap_v2.rs#L55-L57
or you can just add the burn fee part to it instead. This is not fixed yet.

Reply:
> As discussed, burnfee will either stay in contract or sent to an address where these fees will be traded for KMD that will be burned.

Got it. I suggest to send burnFee to burnAddress to not to mix trading payments and burn fees on swap contract address.

> As discussed, total_spend_amount should be used for eth_total_payment

Yep, now agree, if burnFee should be managed by `EtomicSwapTakerV2` then for eth payment I will include burn fee into `eth_total_payment` calculation. 
But it should be added to `eth_total_payment`  when we impl burnFee managing in `EtomicSwapTakerV2`, right now it doesnt have any burnFee functionality https://github.com/KomodoPlatform/etomic-swap/blob/5e15641cbf41766cd5b37b4d71842c270773f788/contracts/EtomicSwapTakerV2.sol

Do you want me to update taker swap contract and provide burnFee field with burnFeeAddress till Wednesday? 
If you would like to refactor smart contract later I suggest not to use `total_spend_amount` as `EtomicSwapTakerV2` doesnt have burnFee support

Reply:
> As discussed, total_spend_amount should be used for eth_total_payment

It means that `dex_fee` will be `fee_amount` + `burn_amount`, but we dont support `burnFee` right now 

![Screenshot 2024-07-29 at 20 55 23](https://github.com/user-attachments/assets/a20f28bc-56b1-4b85-876b-a30210c3b7f2)

Reply:
> But it should be added to eth_total_payment when we impl burnFee managing in EtomicSwapTakerV2, right now it doesnt have any burnFee functionality https://github.com/KomodoPlatform/etomic-swap/blob/5e15641cbf41766cd5b37b4d71842c270773f788/contracts/EtomicSwapTakerV2.sol

It's fine by me to add it then as long as we remember to change it here. I just suggested to use `total_spend_amount` now to not forget to change it later and it will be the same as fee_amount now in defi framework since we didn't add burn fee to EVM TPU in defi framework as well.

> Do you want me to update taker swap contract and provide burnFee field with burnFeeAddress till Wednesday?

No need to, we will add it at the same time we start work on EVM TPU burnfee on defi framework side.

> If you would like to refactor smart contract later I suggest not to use total_spend_amount as EtomicSwapTakerV2 doesnt have burnFee support.
it will lead to the behaviour where we send dex_fee + burn_fee as dexFee to dexFeeAddress (in spendTakerPayment function)

Not really since `burn_fee` will be 0 until we support it in defi framework. And burn fee is taken from the whole fee anyways, we don't let taker pay more so that we can burn KMD, they pay the same total fee with or without burning. But as I answered above, as long as you/we remember to change it later, you can leave it as it is. But `total_spend_amount` was added by @onur-ozkan to use it for total fees whether we burn or not to abstract the burning part, so that other devs can use it without looking into burn fees code.

Reply:
> I just suggested to use total_spend_amount now to not forget to change it later and it will be the same as fee_amount now in defi framework since we didn't add burn fee to EVM TPU in defi framework as well.
>
> But total_spend_amount was added by @onur-ozkan to use it for total fees whether we burn or not to abstract the burning part, so that other devs can use it without looking into burn fees code.

I suppose it depends on the feature. If the swap contract supports only one fee type, then its important to provide the exact fee value in the rust code. Otherwise, the current rust code may not align with the smart contract logic, even if we provide a burn fee of 0.

Its also better to handle fees separately if we agree to provide `burnFee` as a separate field in the smart contract functions.

I added todo sub point in the `[FR] Implement a Fee Splitting and Burning mechanism to burn KMD and support the Komodo blockchain` issue check list https://github.com/KomodoPlatform/komodo-defi-framework/issues/2010#issue-1994776704